internal/tunnel/client/reconnecting.go

```diff
@@ -284,33 +363,79 @@ func (s *reconnectingSession) connect(acceptErr error) error {
 		}
 
 		// dial the tunnel server
-		raw, err := s.dialer()
+		raw, err := s.dialer(connSession.legNumber)
 		if err != nil {
 			failTemp(err, raw)
 			continue
 		}
 
 		// successfully reconnected
-		s.swapper.set(raw)
+		connSession.swapper.set(raw)
 
 		// callback for authentication
-		if err := s.cb(s); err != nil {
+		desiredLegs, err := s.cb(s, raw, connSession.legNumber)
+		if err != nil {
 			failTemp(err, raw)
 			continue
 		}
 
+		// check if more sessions need to be established
+		sendStateChange := true
+		if desiredLegs > len(s.sessions) {
+			// set up the next connection
+			s.createTunnelClientSession(s)
+			// not done with initial setup yet
+			sendStateChange = false
+		}
+
 		// re-establish binds
-		err = restartBinds(raw)
+		err = restartBinds(connSession)
 		if err != nil {
 			failTemp(err, raw)
 			continue
 		}
 
-		// reset wait
-		boff.Reset()
+		if sendStateChange {
+			// reset wait
+			boff.Reset()
 
-		s.Info("client session established")
-		s.stateChanges <- nil
+			s.Info("client session established")
+			s.stateChanges <- nil
+		}
 		return nil
 	}
 }
+
+func reconnectTunnelToSession(raw RawSession, t *tunnel, newTunnels map[string]*tunnel, oldID string) error {
```

Comment:
i feel like this should be a method on the reconnecting session?

Reply:
it didn't need that state so i didn't attach it, but yeah no one else should be calling this, will add 👍